src/peft/tuners/lora/config.py

```diff
@@ -282,6 +309,15 @@ class LoraConfig(PeftConfig):
             )
         },
     )
+    runtime: LoraConfigRuntime = field(default_factory=LoraConfigRuntime, metadata={"help": "Runtime configurations"})
+
+    def to_dict(self):
```

Comment:
I have a small concern about having to use `to_dict`.

PEFT is used as a utility library by other packages. What I'm afraid could happen is that some of these packages may not use `peft_config.save_pretrained` directly but instead call `config = asdict(peft_config)` to do some stuff on the dict before persisting it. Then it would contain the runtime entry. If such a config is later loaded, it would be subtly different to what we would normally get with PEFT.

Maybe it's not a huge concern overall. I tried to think of a way to ensure that `asdict(peft_config)` would also remove the `runtime` key, but that would require some hacking (e.g. popping it from `__dataclass_fields__`) and I'm not sure if that would break more things down the line.

Anyway, if you have a smart idea to resolve this, LMK.

Reply:
I definitely see the concern. I think the best way to address that is to ensure that we don't adhere to runtime configs found in loaded configurations. That way we can nearly guarantee that it won't actually matter.

And I think my tests so far are checking that this is indeed the case. Perhaps hand crafted config files that add runtime_config to them and then load those and checking that it's ignored should be added to the tests too.

Reply:
I agree that would be a good test. I think for that test to pass, we need an update to `from_pretrained` that removes the runtime config if present.

Reply:
You mean in `PeftModel.from_pretrained()`? I was hoping we could avoid that, but I think you're right, as w can't guarantee users won't throw runtime configs in odd places.

Reply:
I dug around a bit and I'm not sure this is needed. I could be missing something, but:

1. The user provides a config which contains a runtime config: the ephemeral GPU offload flag is overridden by the `from_pretrained` code, resulting in a NOOP.
2. The user config does not contain a runtime config: expected.

Reply:
I meant the `PeftConfigMixin.from_pretrained` should be adapted. E.g. this piece of code would currently fail:

```python
import json
from dataclasses import asdict
from peft import LoraConfig, LoraRuntimeConfig

config = LoraConfig(runtime_config=LoraRuntimeConfig(ephemeral_gpu_offload=True))
dct = asdict(config)

assert dct["runtime_config"]["ephemeral_gpu_offload"] is True

with open("/tmp/peft/1859/adapter_config.json", "w") as f:
    json.dump(dct, f)

with open("/tmp/peft/1859/adapter_config.json", "r") as f:
    loaded = json.load(f)

# this fails
assert loaded["runtime_config"]["ephemeral_gpu_offload"] is False
```

Reply:
That code isn't really doing anything inside the PEFT library though, it's just saving/loading JSON data. Since people might (incorrectly) put runtine config stuff in the JSON files themselves, the sane approach is to ignore it *inside PEFT* (and write helpful warnings when encountered), rather than trying to keep it out of the config files.

You could also flip it around: in a near future, we might actually *want* to pass runtime config content around using asdict, e.g. when serializing configurations for distributed training or something. In these cases, the above would be fully expected and desired behavior.

Reply:
Hmm, but I don't see the harm in adding a check to `PeftModel.from_pretrained` and removing the runtime config there. The example code could be considered "outside PEFT", but using `asdict` on a PEFT config and getting a valid dict representation would be a reasonable assumption for a package that builds on top of PEFT.

> in a near future, we might actually _want_ to pass runtime config content around using asdict, e.g. when serializing configurations for distributed training or something

In that case, we can always make further changes. But as I said, I don't even want to change the output of `asdict(config)` but rather just ensure that when we read the config, we don't accidentally include the runtime config.

Reply:
OK, I think I got a good solution. How does 40ff66b461437d4ea171ab6d337fa2954a286afd look to you?

Reply:
OK, I think I got a good solution. How does 40ff66b461437d4ea171ab6d337fa2954a286afd look to you?